### ISSUE TITLE
Incompatible es6 depedencies

### ISSUE THREAD

**Issue by joshq00** [google/material-design-lite#2032]
I am failing to replicate the drone build on my end.
```sh
$ docker run -ti crhym3/ci-image /bin/bash
root@473e5c233846:/go# mkdir /workspace
root@473e5c233846:/go# cd /workspace
root@473e5c233846:/workspace# git clone https://github.com/google/material-design-lite.git --depth 1
root@473e5c233846:/workspace/material-design-lite# npm i -qq
root@473e5c233846:/workspace/material-design-lite# gulp all
```
> ```
[16:54:52] Requiring external module babel-core/register
/workspace/material-design-lite/node_modules/gulp-closure-compiler/node_modules/google-closure-compiler/lib/gulp/index.js:106
          let outputFiles;
          ^^^
SyntaxError: Unexpected strict mode reserved word
    at exports.runInThisContext (vm.js:73:16)
    at Module._compile (module.js:443:25)
    at Module._extensions..js (module.js:478:10)
    at Object.require.extensions.(anonymous function) [as .js] (/workspace/material-design-lite/node_modules/babel-core/lib/api/register/node.js:214:7)
    at Module.load (module.js:355:32)
    at Function.Module._load (module.js:310:12)
    at Module.require (module.js:365:17)
    at require (module.js:384:17)
    at Object.<anonymous> (/workspace/material-design-lite/node_modules/gulp-closure-compiler/node_modules/google-closure-compiler/index.js:26:19)
    at Module._compile (module.js:460:26)
```

**Comment by joshq00** [comment]
Some info:
```sh
$ docker images
REPOSITORY            TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
crhym3/ci-image       latest              998016c3e12b        12 days ago         1.52 GB
```
This matches the newest image in [crhym3/ci-image](https://hub.docker.com/r/crhym3/ci-image/)

```sh
root@473e5c233846:/go# node version
v0.12.8
```

```sh
root@473e5c233846:/workspace/material-design-lite# git log --oneline -1
3d66281 Merge pull request #2029 from listepo/gitter
```

**Comment by samccone** [comment]
@joshq00 this is an es6 issue, I am guessing drone is on node 0.12 .. and let was not added to v8 until node4 

node 0.12 is the new IE

**Comment by samccone** [comment]
ref: https://github.com/google/material-design-lite/issues/1916

**Comment by samccone** [comment]
cc @crhym3

**Comment by joshq00** [comment]
Haha agreed on 0.12. 

Changing `.drone.yml` to use node:5.2.0 (instead of the ci-image) gets gulp all to work, but mocha fails. Don't know which is the lesser of the two evils

**Comment by samccone** [comment]
welcome to javascript in 2016 :fire: :imp:

**Comment by joshq00** [comment]
I'm confused how it works on your end. Maybe something with the cached `node_modules`? Using a different docker image?

**Comment by crhym3** [comment]
Re: replicating drone build. The only thing that comes to mind is cache. I made drone cache `node_modules`, otherwise it takes forever.

Re: node 0.12. That's exactly the version [it claims to be compatible with in the readme](https://github.com/google/material-design-lite#build).

**Comment by crhym3** [comment]
So, what is this issue about exactly, min compatible node version or replicating a drone build?

**Comment by samccone** [comment]
oh yeah it is a cache issue then, since prob one of the dependencies updated their codebases (inside of the semver range) to use new es6 features... but then broke everything :crying_cat_face:

**Comment by joshq00** [comment]
Good point, @crhym3: does mdl only *appear* to be compatible with 0.12.8 due to cached dependencies?

**Comment by joshq00** [comment]
Maybe a stricter set of semver options in package.json can resolve it...

**Comment by samccone** [comment]
shrinkwrap.json would be the real fix :)

@crhym3 can you run npm shrinkwrap on the server and paste the file somewhere?

**Comment by joshq00** [comment]
But also, is there a max compatible version? As I mentioned, the tests can't run on 5.2.0

**Comment by crhym3** [comment]
I've just purged the cache and restarted latest drone build: https://ci.cloudware.io/google/material-design-lite/77

Let's see what happens.

**Comment by crhym3** [comment]
Yeah, it fails with the same error mentioned in https://github.com/google/material-design-lite/issues/2032#issue-123519875

Any new PR or commit to an existing PR will now fail. Should be fairly easy to repro and fix.

@samccone sorry I have no idea what shrinkwrap is and I try not to go into node and npm deps as much as I can.

**Comment by crhym3** [comment]
But anyone can do it, really. Just use instructions posted by @joshq00 in https://github.com/google/material-design-lite/issues/2032#issue-123519875

**Comment by crhym3** [comment]
Alternatively, just use node v0.12, `rm -rf node_modules && npm i && gulp all`. You should be able to repro w/o drone/docker container.

**Comment by joshq00** [comment]
To prevent this in the future – and continue using cache for build speed – a step for `npm update` could be added

**Comment by crhym3** [comment]
@joshq00 do you mean something like this?

```yaml
commands:
  - npm install
  - npm update
  - ...
```

**Comment by crhym3** [comment]
Wouldn't be sufficient to just replace `npm install` with `npm update`?

**Comment by joshq00** [comment]
@crhym3 yes, exactly.
And I believe you're right.

However, [the docs](https://docs.npmjs.com/cli/update) mention
> As of npm@2.6.1, the npm update will only inspect top-level packages. Prior versions of npm would also recursively inspect all dependencies
> ...
> be warned that simultaneous asynchronous update of all packages, including npm itself and packages that npm depends on, often causes problems up to and including the uninstallation of npm itself.

**Comment by crhym3** [comment]
Yeah, no kidding package manager. Anyway, let's try with `npm update` for now and see how it goes.

**Comment by crhym3** [comment]
Ok, I update drone script to run `npm update` instead of `npm install` in https://github.com/google/material-design-lite/commit/1a28d77e1519c8804aee33fc8904b84f846eb5de.

**Comment by crhym3** [comment]
@samccone re: #1916. what would moving to travis solve in this case?

**Comment by samccone** [comment]
oh just dropping the complexity a bit, travis takes care of smart caching deps, and would allow us to test a range of node versions, ideally catching this ^  issue a bit earlier :)

**Comment by crhym3** [comment]
Oh, I absolutely disagree. In fact, Drone caching is more secure than Travis. Just read the docs on both, you'll see the difference.

Drone also has matrix builds, identical to Travis, and it is as easy to setup.

In fact, this very issue shows how much easier it is to repro and identify a build issue, well dependency in this case, using docker containers drone is based on. It is technically not possible to repro Travis builds like @joshq00 did in https://github.com/google/material-design-lite/issues/2032#issue-123519875.

**Comment by samccone** [comment]
> @samccone sorry I have no idea what shrinkwrap is and I try not to go into node and npm deps as much as I can.

Prior to flushing the cache on the docker container, if you were to run npm shrinkwrap it would have given us an exact snapshot of the version of all of the node_modules installed. Taking the shrinkwrap file and then adding it to the repo would make it so anyone running `npm i` would get a working version of all of the node_modules... To be clear the goal I had in mind was to make sure the repo was in a passing state (from the shrinkwrap.json file) while someone bisected the deps and found which one broke things. 

As far as why Drone is better, I need to do some more reading I guess, I am only trying to lower to burden of upkeep for you, and make it easier for others on the team to update / tweak the build && test phase (in my mind travis would give us this, however I totally could be misinformed here... I will do more research)

**Comment by crhym3** [comment]
Re: shrinkwrap, isn't it already seen from the error msg?

```
/workspace/material-design-lite/node_modules/gulp-closure-compiler/node_modules/google-closure-compiler/lib/gulp/index.js:106
          let outputFiles;
          ^^^
```

Based on the error msg and the fact that `gulp-closure-compiler` has [no version specified in package.json](https://github.com/google/material-design-lite/blob/1a28d77e1519c8804aee33fc8904b84f846eb5de/package.json#L23), I'd make a wild guess it's `gulp-closure-compiler`, no? Looking at [their commits](https://github.com/steida/gulp-closure-compiler/commits/master), there seems to a fairly recent version bump. I'd just try the previous one.

FYI, there's also [google-closure-compiler in package.json w/o a version](https://github.com/google/material-design-lite/blob/1a28d77e1519c8804aee33fc8904b84f846eb5de/package.json#L19).

**Comment by sgomes** [comment]
This should be fixed in efefcfc7eb6b1a69d67e4b2b134d6b3ffd503b92, thanks for the issue @joshq00!
